Store the QueryBuilder's Writable representation instead of its XContent representation

martijnvg opened the pull request:
The Writeble representation is less heavy to parse and that will benefit percolate performance and throughput. The performance gain of this change depends on the complexity of the stored percolator queries. A percolator query with only simple term query will not benefit as much as bool query with many clauses. A simple test with percolator queries having bool query with two should clauses showed a 35% improvement in query time. Percolator queries do need to be re-indexed in order to benefit from this change.

The change looks bigger than it is. The main change is in `PercolatorFieldMapper` and `PercolateQueryBuilder`. The changes in the other files are related to make sure that `NamedWriteableRegistry` is available in the `PercolateQueryBuilder`.

jpountz commented:
I think this is a better representation indeed. In terms of backward compat, I think we have several options:
 - extend the compatibility guarantees for the binary representation of query builders
 - store both the xcontent and binary representation of the query and use the binary representation whenever possible and the xcontent as a fallback
 - only store the binary representation of the query in doc values and fall back to the `_source` if the index creation version is earlier than the min compatibility version of the node?

martijnvg commented:
@jpountz Thanks for looking! I'll label the pr with discuss label to discuss this change with a wider audience.

martijnvg commented:
I removed the discuss label. At least from 6.0, the binary format has the same the same bwc guarantees as the xcontent format. So there this PR doesn't have make trace bwc guarantees for better performance .

martijnvg commented:
My previous statement is incorrect. The binary format is backwards compatible with the latest minor release of the previous major release line and that is different than the backwards compatibility of the xcontent format.

I've updated the PR to take this into account, but instead of adding a fallback parsing mechanism to parse to _source or remain to store the xcontent next to the binary format, the pr now fails to execute the percolate query and index percolator queries if the created index version of the index the percolator queries reside in is before the latest minor release of the previous release line. I believe that this behavior is better than if percolating becomes slower after an upgrading without any clear warning.

If percolator queries are no longer compatible or if upgrading to the next version means that percolator queries are not going to be compatible then percolator indices should re-indexed using the reindex api. This shouldn't be a problem for indices only containing percolator queries, because the number of documents tends to be much lower compared to indices with actual data.

frankkoornstra commented:
Don't know too much about the internals but from a user perspective a performance improvement in percolators would be awesome! Just speaking from my use case: percolators are a projection from another source of truth so backwards compatibility is a non-issue for me. We would just remove the percolators index and re-create it. Thanks for looking into this!

martijnvg commented:
I've updated the pr. Removed the checks that prohibits using the `percolate` query on a too old index and indexing percolator queries into an too old index. Instead the binary representation of query builders is extended to be the same as the xcontent representation. To ensure this a qa test has been added that writes percolator queries with an older version and then reads them with the current version. Currently the qa test only tests against the current version, because there isn't a release that included the changes in this pr. I'm going to test more query builders (currently 2 are tested), but just wanted to check what others think about the testing approach.

jpountz commented on qa/query-builder-bwc/src/test/java/org/elasticsearch/bwc/QueryBuilderBWCIT.java:
```diff
@@ -0,0 +1,220 @@
+                byte[] qbSource = Base64.getDecoder().decode(queryBuilderStr);
+                try (InputStream in = new ByteArrayInputStream(qbSource, 0, qbSource.length)) {
+                    try (StreamInput input = new NamedWriteableAwareStreamInput(new InputStreamStreamInput(in), registry)) {
+                        input.setVersion(Version.CURRENT);
```
why CURRENT? it should be the index created version?

martijnvg replied:
good catch!

jpountz commented on qa/query-builder-bwc/src/test/java/org/elasticsearch/bwc/QueryBuilderBWCIT.java:
```diff
@@ -0,0 +1,220 @@
+                    try (StreamInput input = new NamedWriteableAwareStreamInput(new InputStreamStreamInput(in), registry)) {
+                        input.setVersion(Version.CURRENT);
+                        QueryBuilder queryBuilder = input.readNamedWriteable(QueryBuilder.class);
+                        assertEquals(expectedQueryBuilder, queryBuilder);
```
let's also make sure that we read all the bytes?

martijnvg commented on docs/reference/mapping/types/percolator.asciidoc:
```diff
@@ -62,6 +62,181 @@ put index and put mapping APIs.
+
+Percolator queries have a limited backwards compatibility between major versions. A node is able to read percolator
+queries from an index created with up to the latest minor version of the previous major release line. A node is always
+able to read percolator queries between minor versions of the current major release.
```
left over...

martijnvg commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java:
```diff
@@ -575,32 +581,55 @@ public Weight createNormalizedWeight(Query query, boolean needsScores) throws IO
+                SortedBinaryDocValues values = context.getForField(queryBuilderFieldType).load(ctx).getBytesValues();
+                return docId -> {
+                    if (values.advanceExact(docId)) {
+                        BytesRef qbSource = values.nextValue();
```
So this change is a side effect of making the the query builder field inside the percolator field accessible via doc values fields. I needed that in order to make the query builder bwc qa test to work. 

When field values via doc values fields it uses `ScriptDocValues` to access doc values. (DocValueFieldsFetchSubPhase.java line 88) This uses a special encoding to deal with multi values binary values.

martijnvg replied:
So I changed this in the latest commit to now not use our field data apis, but `BinaryDocValues` directly.

martijnvg commented on modules/percolator/src/test/java/org/elasticsearch/percolator/PercolatorFieldMapperTests.java:
```diff
@@ -480,9 +489,14 @@ public void testUnsupportedQueries() {
+        try (InputStream in = new ByteArrayInputStream(actual.bytes, actual.offset, actual.length)) {
+            try (StreamInput input = new NamedWriteableAwareStreamInput(new InputStreamStreamInput(in), writableRegistry())) {
+                input.readVInt();
+                input.readVInt();
```
I will add a comment. The reason two vints are read is because of the format for binary fields.

martijnvg commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java:
```diff
@@ -575,32 +581,55 @@ public Weight createNormalizedWeight(Query query, boolean needsScores) throws IO
+                                    QueryBuilder queryBuilder = input.readNamedWriteable(QueryBuilder.class);
+                                    return PercolatorFieldMapper.toQuery(context, mapUnmappedFieldsAsString, queryBuilder);
+                                }
+                            }
```
Maybe I'm not seeing it, but the input classes here don't have some kind of a reset method.

jpountz commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java:
```diff
@@ -575,32 +582,63 @@ public Weight createNormalizedWeight(Query query, boolean needsScores) throws IO
+                return docId -> {
+                    if (binaryDocValues.advanceExact(docId)) {
+                        BytesRef qbSource = binaryDocValues.binaryValue();
+                        if (qbSource.length > 0) {
```
this check should not be necessary?

martijnvg replied:
now I think about it: no :)

martijnvg commented:
@jpountz I labelled this issue backport pending in order to discuss whether this should be backported to 6.0.

martijnvg commented:
We are going to backport this change after 6.0.0-beta1 has been released.

martijnvg commented:
This has been backported to the 6.0 branch: 77c73c29c1a276bb70c55c3fd0bbaf9fce54c032
